src/parser.rs

```diff
@@ -333,6 +333,12 @@ pub fn copy_token(ptr: TokenPtr) -> Token {
     ptr.to_owned().into_boxed_slice()
 }
 
+/// Transmutes this token into a Rust string. This is safe for us
+/// because we check that our tokens are valid ASCII (and thus also valid UTF8).
+pub fn as_str(ptr: TokenPtr) -> &str {
+    unsafe { str::from_utf8_unchecked(ptr) }
```

Comment:
Do you have benchmarks showing that the utf-8 check consumes more than 1% of runtime under a plausible workload?  I think the burden of proof should be on whoever is adding a new unsafe block, *especially* if that unsafe block doesn't have a clear abstraction barrier and effectively depends on everything else in the program for correctness.